src/services/codefixes/inferFromUsage.ts

```diff
@@ -340,6 +376,27 @@ namespace ts.codefix {
         }
     }
 
+    function inferTypeForThisFromUsage(containingFunction: SignatureDeclaration, program: Program, cancellationToken: CancellationToken) {

```

Comment:
can you use the existing entrypoint inferTypeForParametersFromUsage/Reference?

I think you can get by with (1) an additional boolean "look for thisType" (2) checks that containingFunction.kind must be FunctionExpression, FunctionDeclaration or MethodDeclaration.

Reply:
Sorry, I don't follow 🙁.  inferTypeForParametersFromUsage directly returns one array now. Do you mean you'd like me to change it to return an object with an array and/or thisType?

Reply:
Re-reading the code, the part I don't like is that `inferTypeForThisFromUsage`'s code is a subset of `inferTypesForParmaetersFromUsage`, which makes sense since `this` is an implicit parameter. `inferTypeForThisFromReferences` is the same, although it's smaller.

I think you can 
1. pass inferTypeForParmetersFromUsage + inferTypeForParametersFromReferences an extra boolean to make it infer `this` instead of parameters, then return a one-entry array. Then you can call inferTypeForPararametersFromUsage from annotateThis. When run this way, you need some assertions that containingFunction.kind is one of the legal kinds for `this` parameters.
-OR-
2.  Extract the shared code of inferTypesForParmaetersFromUsage and inferTypeForThisFromusage into its own function.

I prefer (1) since I think `this` and parameters are conceptually similar and it's likely they will share more code in the future.

Reply:
Got it - that makes sense. Thanks for the explanation.

`inferTypeForParametersFromReferences` returns a `ParameterInference[]` with a `declaration: ParameterDeclaration`, but that doesn't seem relevant for a `this` inference. So I ended up going with (2): `inferFunctionReferencesFromUsage`. Sorry to go against your preference 😄.